subprojects/core/src/main/java/org/gradle/initialization/ProjectPropertySettingBuildLoader.java

```diff
@@ -22,139 +22,140 @@
+        Map<String, String> resolvedProjectProperties = gradleProperties.mergeProperties(intermediateProjectProperties);
 
-    /**
-     * Applies the given properties to the project and its subprojects, caching property mutators whenever possible
-     * to avoid too many searches.
-     */
-    private static class CachingPropertyApplicator {
-        private final Class<? extends Project> projectClass;
-        private final Map<Pair<String, ? extends Class<?>>, Optional<PropertyMutator>> mutators = new HashMap<>();
-        private ImmutableMap.Builder<String, Object> extraProjectProperties;
+        ProjectMutatingProperties mutatingProperties = new ProjectMutatingProperties();
+        ImmutableMap.Builder<String, String> extraProjectProperties = ImmutableMap.builder();
 
-        CachingPropertyApplicator(Class<? extends Project> projectClass) {
-            this.projectClass = projectClass;
+        for (Map.Entry<String, String> entry : resolvedProjectProperties.entrySet()) {
+            attributeProperty(mutatingProperties, extraProjectProperties, entry.getKey(), entry.getValue());
         }
 
-        void configureProperty(Project project, String name, Object value) {
-            if (isPossibleProperty(name)) {
-                assert project.getClass() == projectClass;
-                PropertyMutator propertyMutator = propertyMutatorFor(name, typeOf(value));
-                if (propertyMutator != null) {
-                    propertyMutator.setValue(project, value);
-                } else {
-                    extraProjectProperties.put(name, value);
-                }
-            }
-        }
+        return new ResolvedProjectProperties(mutatingProperties, extraProjectProperties.build());
+    }
 
-        @Nullable
-        private static Class<?> typeOf(@Nullable Object value) {
-            return value == null ? null : value.getClass();
+    private static void attributeProperty(
+        ProjectMutatingProperties mutatingProperties,
+        ImmutableMap.Builder<String, String> extraProjectProperties,
+        String propertyName,
+        String propertyValue
+    ) {
+        if (propertyName.isEmpty()) {
+            // Historically, we filtered out properties with empty names here.
+            // They could appear in case the properties file has lines containing only '=' or ':'
+            return;
         }
 
-        @Nullable
-        private PropertyMutator propertyMutatorFor(String propertyName, @Nullable Class<?> valueType) {
-            return mutators.computeIfAbsent(
-                Pair.of(propertyName, valueType),
-                key -> {
-                    assert key.left != null;
-                    return Optional.ofNullable(writeablePropertyIfExists(projectClass, key.left, key.right));
-                }
-            ).orElse(null);
+        switch (propertyName) {
+            case "version":
+                mutatingProperties.version = propertyValue;
```

Comment:
❓ It is a good practice to separate side-effects from computation, in this case though, it seems we could eliminate the `ProjectMutatingProjects` abstraction and mutate the project properties directly in this switch-case without losing much/anything. Your thoughts?